tracing-serde/src/lib.rs

```diff
@@ -156,20 +186,26 @@ impl<'a> Serialize for SerializeRecord<'a> {
         S: Serializer,
     {
         let serializer = serializer.serialize_map(None)?;
-        let mut visitor = SerdeMapVisitor {
-            serializer,
-            state: Ok(()),
-        };
+        let mut visitor = SerdeMapVisitor::new(serializer, Ok(()));
         self.0.record(&mut visitor);
         visitor.finish()
     }
 }
 
-struct SerdeMapVisitor<S: SerializeMap> {
+pub struct SerdeMapVisitor<S: SerializeMap> {
     serializer: S,
     state: Result<(), S::Error>,
 }
 
+impl<S> SerdeMapVisitor<S>
+where
+    S: SerializeMap,
+{
+    pub fn new(serializer: S, state: Result<(), S::Error>) -> Self {
```

Comment:
The `new` function should probably assume the `state` is `Ok`, since we haven't written anything to the serializer yet. So what about
```rust
    pub fn new(serializer: S) -> Self {
        Self { serializer, state: Ok(()) }
    }
```

Reply:
Having some annoyances about the type signatures if I do it this way. Actually tried it first, but using this elsewhere was super annoying. Except I don't get these anymore so I guess rustc can infer the types now.